### PR TITLE
Windows version for Focus-Point plugin

### LINKED ISSUE #70: Focus Points Plugin clips on Windows
The plugin mostly works on windows, minus this clipping issue (see screenshot). I'm not sure there's any way around this other than using imagemagick to bypass Lightroom rendering. 

screen

### PR DESCRIPTION
**Pull request by ropma**
- enhance InfoProvider with option section for mogrify
- DefaultPointRenderer.createView returns the final photo view including overlays / mogrify frames
- the old implemenation of "createView" is splitted into two new methods
  - prepareRendering: rendering independent  part
  - createOverlayViews: create overlay views
- new helper class MogrifyUtils for mogrify handling (export to disk, create focus frames)

closes #70

### CODE REVIEW

**Comment by musselwhizzle** [comment]
FYI, i changed this into a branch called "windows_support" so that i can make sure all the mac stuff still works. I'll merge those two branches once I can QA.

File: README.md [hunk 1]
```diff
@@ -34,11 +34,18 @@ Supported Cameras
 
 Installing
 --------
+**Prerequisite:**
```
Comment (musselwhizzle) [inline README.md]: Can we just include `mogrify.exe` in the bin folder and dynamically get the path to it so user's don't have to do this extra step?

Comment (ropma) [reply README.md]: I just check the license.txt of ImageMagick (mogrify) and it seems to be possible to include it to the focus-point distribution:
---
It allows you to:

  * freely download and use ImageMagick software, in whole or in part, for personal, company internal, or commercial purposes;
  * use ImageMagick software in packages or distributions that you create;
  * link against a library under a different license;
  * link code under a different license against a library under this license;
  * merge code into a work under a different license;
  * extend patent grants to any code using code under this license;
  * and extend patent protection.

It forbids you to:

  * redistribute any piece of ImageMagick-originated software without proper attribution;
  * use any marks owned by ImageMagick Studio LLC in any way that might state or imply that ImageMagick Studio LLC endorses your distribution;
  * use any marks owned by ImageMagick Studio LLC in any way that might state or imply that you created the ImageMagick software in question.

It requires you to:

  * include a copy of the license in any redistribution you may make that includes ImageMagick software;
  * provide clear attribution to ImageMagick Studio LLC for any distributions that include ImageMagick software.

It does not require you to:

  * include the source of the ImageMagick software itself, or of any modifications you may have made to it, in any redistribution you may assemble that includes it;
  * submit changes that you make to the software back to the ImageMagick Studio LLC (though such feedback is encouraged).

File: focuspoints.lrdevplugin/MogrifyUtils.lua [hunk 2]
```diff
@@ -0,0 +1,181 @@
+--[[
+
+MIT License
```
[the remark this answers is not in the document]
Comment (ropma) [reply focuspoints.lrdevplugin/MogrifyUtils.lua]: Oh yes, of cause. I copy the file from the prototype project I did and forgot to change the license type.

File: README.md (hunk 1 above)
Comment (ropma) [reply README.md]: I can try to do this tomorrow...

Comment (ropma) [reply README.md]: Done. I've committed the following changes in ropma/windows-support:

Windows Support - add ImageMagick to plugin distribution
- add ImageMagick to bin folder
- InfoProvider: remove Options section (not needed anymore as mogrify is part of the plugin distribution)
- InfoProvider: add Acknowledgements section
- change license type of MogrifyUtils.lua
- update README.md

Test on a "clean" windows box was successful, thus the ImageMagick installation in the plugin bin folder is self contained.

Comment (musselwhizzle) [reply README.md]: I will leave this PR up for a day or 2 incase there is anything you want to change and then will merge it in to the branch. I'll give a QA pass on Mac and then merge it over. Once I tell people it works on Windows, you're probably going to get a huge influx of people using it (and probably giving feedback on how it works). Thanks for your effort on this. Couldn't have done it without you!

Comment (ropma) [reply README.md]: Also thanks to you for creating and maintaining this project!! I really enjoy to use the plugin.

Great if you can do a QA on Mac. I don't have one and only did some basic tests with the  overlay mode after refactoring.

**Comment by capricorn8** [comment]
Really really looking forward to using this as well. As soon something is available to test on windows I'm ready to do so :-)

**Comment by musselwhizzle** [comment]
Screen Shot 2019-10-02 at 4 08 19 PM
Error on mac. haven't looked into it yet

**Comment by capricorn8** [comment]
Great job !!! 👍👍

Works like a charme for my Olympus ORF files on Win10, with one exception. For some files I get an
image
Digging deeper, it seems to me that for these files there is a problem with the rectangular definition for mogrify. When I execute the command given in the logfile on command line 

_>magick.exe mogrify -fill none -strokewidth 3 -stroke red -draw "roundRectangle 434,524 335,591 1,1" C:\Temp\P4001793-fpoints.jpg_

I get the following:

_mogrify: non-conforming drawing primitive definition `roundRectangle' @ error/draw.c/RenderMVGContent/4396._

**Comment by ropma** [comment]
When I execute the command printed in the logfile on command line
> 
> _>magick.exe mogrify -fill none -strokewidth 3 -stroke red -draw "roundRectangle 434,524 335,591 1,1" C:\Temp\P4001793-fpoints.jpg
> 
> I get the following:
> 
> _mogrify: non-conforming drawing primitive definition `roundRectangle' @ error/draw.c/RenderMVGContent/4396._
> 
> See attached file to reproduce the issue.
> 
> [P4001793.zip](https://github.com/musselwhizzle/Focus-Points/files/3685079/P4001793.zip)



> Great job !!! 👍👍
> 
> Works like a charme for my Olympus ORF files on Win10, with one exception. For some files I get an
> image
> Digging deeper, it seems that for these files there is a problem with improper input to mogrify. When I execute the command printed in the logfile on command line
> 
> _>magick.exe mogrify -fill none -strokewidth 3 -stroke red -draw "roundRectangle 434,524 335,591 1,1" C:\Temp\P4001793-fpoints.jpg
> 
> I get the following:
> 
> _mogrify: non-conforming drawing primitive definition `roundRectangle' @ error/draw.c/RenderMVGContent/4396._
> 
> See attached file to reproduce the issue.
> 
> [P4001793.zip](https://github.com/musselwhizzle/Focus-Points/files/3685079/P4001793.zip)

Thank you for testing and the feedback. I'll checked the file in the attachment and couldn't reproduce the problem. The focus point is shown at the head.
Which windows version are you using?

**Comment by ropma** [comment]
> Screen Shot 2019-10-02 at 4 08 19 PM
> 
> Error on mac. haven't looked into it yet

That's strange. Does the Mac version try to use mogrify? Before I added mogrify to the distribution is was also possible on a Mac to use mogrify. This was handled by a prefs param which was set in the Option dialog of the info provider. This parameter still exist. I'll make an update, maybe that's the root case of the problem.

**Comment by capricorn8** [comment]
> Thank you for testing and the feedback. I'll checked the file in the attachment and couldn't reproduce the problem. The focus point is shown at the head.

That is strange indeed. For this, and few other files the error is clearly reproducible.

> Which windows version are you using?

Lightroom Classic-Version: 8.4.1 [ 201909111355-eb9b68f0 ]
Lizenz: Creative Cloud
Spracheinstellung: de
Betriebssystem: Windows 10 - Business Edition
Version: 10.0.18362
Anwendungsarchitektur: x64

**Comment by capricorn8** [comment]
I guess I found the reason. 

Looking at how the parameters for "draw roundRectangle" are composed in your code, the logic of top-left and bottom-right coordinates does not match with the actual figures in the logfile above, where this is exactly reversed: (434,521)=br, (335,591)=tl.

And indeed, those images where the problem occurs I have been flipped horizontally in LR with regard to their original orientation.

Are you able to you detect / handle this?

**Comment by ropma** [comment]
Thanks for the analysis. I'll check this.

**Pull request merged**